lib/DXIL/DxilOperations.cpp

```diff
@@ -3249,23 +3249,24 @@ void OP::GetMinShaderModelAndMask(const llvm::CallInst *CI,
   if (DXIL::CompareVersions(valMajor, valMinor, 1, 5) < 0) {
     // validator 1.4 didn't exclude wave ops in mask
     if (IsDxilOpWave(opcode))
-      mask = ((unsigned)1 << (unsigned)DXIL::ShaderKind::Invalid) - 1;
-    // These shader models don't exist before 1.5
-    mask &= ~(SFLAG(Amplification) | SFLAG(Mesh));
+      mask = ((unsigned)1 << (unsigned)DXIL::ShaderKind::Mesh) - 1;
     // validator 1.4 didn't have any additional rules applied:
     return;
   }
 
   // Additional rules are applied manually here.
 
-  // Barrier with mode != UAVFenceGlobal requires compute, amplification, or
-  // mesh Instructions: Barrier=80
+  // Barrier with mode != UAVFenceGlobal requires compute, amplification,
+  // mesh, or node. Instructions: Barrier=80
   if (opcode == DXIL::OpCode::Barrier) {
-    DxilInst_Barrier barrier(const_cast<CallInst *>(CI));
-    unsigned mode = barrier.get_barrierMode_val();
-    if (mode != (unsigned)DXIL::BarrierMode::UAVFenceGlobal) {
-      mask =
-          SFLAG(Library) | SFLAG(Compute) | SFLAG(Amplification) | SFLAG(Mesh);
+    // Barrier mode should be a constant, but be robust to non-constants here.
+    if (isa<ConstantInt>(CI->getArgOperand(DxilInst_Barrier::arg_barrierMode))) {
+      DxilInst_Barrier barrier(const_cast<CallInst *>(CI));
+      unsigned mode = barrier.get_barrierMode_val();
+      if (mode != (unsigned)DXIL::BarrierMode::UAVFenceGlobal) {
+        mask &= SFLAG(Library) | SFLAG(Compute) | SFLAG(Amplification) |
```

Comment:
Just double checking here.  The original code just assigned an OR'd set of flags directly 'mask = flag | flag | flag'.  This new logic is ANDing this set of bits now. 'mask &= flag | flag | flag'.  Is this expected?

Reply:
I believe the `&=` is a safer pattern to use whenever further restricting the mask based on any special case following the `GetMinShaderModelAndMask` call, which initializes the mask and should set it based on specific instruction requirements.

Special cases like this need to restrict the mask, not accidentally expand it.  The generated code for the DXIL op should have set the maximally permissive mask for the operation, which we might need to restrict with some special cases here.

Assignment would have been equivalent in this case, but it's not a pattern this code should use unless we are trying to match a prior validator in a special case for compatibility, like the earlier code does for wave ops.

However, this comment made take a second look at another case in this function, and I realized it needed updating, so thanks for that!  I'll update that in a little while after I can craft and verify a test that hits that case.